blockchain/chain_sync/src/tipset_syncer.rs

```diff
@@ -0,0 +1,1741 @@
+use std::cmp::min;
+use std::collections::{HashMap, HashSet};
+use std::convert::TryFrom;
+use std::error::Error as StdError;
+use std::marker::PhantomData;
+use std::sync::Arc;
+use std::time::{SystemTime, UNIX_EPOCH};
+
+use async_std::channel::Receiver;
+use async_std::future::Future;
+use async_std::pin::Pin;
+use async_std::stream::{Stream, StreamExt};
+use async_std::task::{self, Context, JoinHandle, Poll};
+use futures::stream::FuturesUnordered;
+use log::{debug, error, info, trace, warn};
+use num_bigint::BigInt;
+use thiserror::Error;
+
+use crate::bad_block_cache::BadBlockCache;
+use crate::network_context::SyncNetworkContext;
+use crate::sync_state::{SyncStage, SyncState};
+use crate::validation::TipsetValidator;
+use actor::{is_account_actor, power};
+use address::Address;
+use beacon::{Beacon, BeaconEntry, BeaconSchedule, IGNORE_DRAND_VAR};
+use blocks::{Block, BlockHeader, Error as ForestBlockError, FullTipset, Tipset, TipsetKeys};
+use chain::Error as ChainStoreError;
+use chain::{persist_objects, ChainStore};
+use cid::Cid;
+use clock::ChainEpoch;
+use crypto::{verify_bls_aggregate, DomainSeparationTag, Signature};
+use encoding::Cbor;
+use encoding::Error as ForestEncodingError;
+use fil_types::{
+    verifier::ProofVerifier, NetworkVersion, Randomness, ALLOWABLE_CLOCK_DRIFT, BLOCK_GAS_LIMIT,
+    TICKET_RANDOMNESS_LOOKBACK,
+};
+use forest_libp2p::chain_exchange::TipsetBundle;
+use interpreter::price_list_by_epoch;
+use ipld_blockstore::BlockStore;
+use message::{Message, UnsignedMessage};
+use networks::{get_network_version_default, BLOCK_DELAY_SECS, UPGRADE_SMOKE_HEIGHT};
+use state_manager::Error as StateManagerError;
+use state_manager::StateManager;
+use state_tree::StateTree;
+
+const MAX_TIPSETS_TO_REQUEST: u64 = 100;
+
+#[derive(Debug, Error)]
+pub enum TipsetProcessorError {
+    #[error("TipsetRangeSyncer error: {0}")]
+    TipsetRangeSyncer(#[from] TipsetRangeSyncerError),
+    #[error("Tipset stream closed")]
+    TipsetStreamClosed,
+    #[error("Tipset has already been synced")]
+    TipsetAlreadySynced,
+}
+
+#[derive(Debug, Error)]
+pub enum TipsetRangeSyncerError {
```

Comment:
Should we put all error defs in a error.rs file for this crate as other modules have a few error defs as well?

Reply:
Breaking it out into a separate file makes sense if it's used by other files, I think. It's always a good rule in Rust to flatten the dependency tree.

Reply:
I like the Elm project's explanation of this code dynamic:
https://guide.elm-lang.org/webapps/structure.html#defensive-instincts

Reply:
All of these error types are internal to the crate, only modules within this crate depend on these errors types. I find it easier to keep them updated when the errors are defined with the types that returns them.